zh/advanced-statefulset.md

```diff
@@ -0,0 +1,156 @@
+---
+title: 使用增强型 StatefulSet 控制器
+summary: 介绍如何开启、使用增强型 StatefulSet 控制器
+category: alpha
+---
+
+# 使用增强型 StatefulSet 控制器
+
+Kubernetes 内置 [StatefulSet](https://kubernetes.io/docs/concepts/workloads/controllers/statefulset/) 为 Pods 分配连续的序号。比如 3 个副本时，Pods 分别为 pod-0, pod-1, pod-2 。扩缩容时，必须在尾部增加或删除 Pods 。比如扩容到 4 个副本时，会新增 pod-3 。缩容到 2 副本时，会删除 pod-2 。 
+
+在使用本地存储时，Pods 与 Nodes 存储资源绑定，无法自由调度。若希望删除掉中间某个 Pod ，以便维护其所在的 Node 但并没有其他 Node 可以迁移时，或者某个 Pod 故障想直接删除，另起一个序号不一样的 Pod 时，无法通过内置 StatefulSet 实现。
+
+[增强型 StatefulSet 控制器](https://github.com/pingcap/advanced-statefulset) 基于内置 StatefulSet 实现，新增了自由控制 Pods 序号的功能。本文介绍如何在 TiDB Operator 中使用。
+
+## 开启
+
+1. 载入 Advanced StatefulSet 的 CRD 文件：
+
+    Kubernetes 1.16 之前版本：
+
+    {{< copyable "shell-regular" >}}
+
+    ```shell
+    kubectl apply -f https://raw.githubusercontent.com/pingcap/tidb-operator/master/manifests/advanced-statefulset-crd.v1beta1.yaml
+    ```
+
+    Kubernetes 1.16 之后:
+
+    {{< copyable "shell-regular" >}}
+
+    ```
+    kubectl apply -f https://raw.githubusercontent.com/pingcap/tidb-operator/master/manifests/advanced-statefulset-crd.v1.yaml
+    ```
+
+2. 在 TiDB Operator chart 的 values 中启用 `AdvancedStatefulSet` 特性：
+
+    {{< copyable "shell-regular" >}}
+
+    ```yaml
+    features:
+    - AdvancedStatefulSet=true
+    advancedStatefulset:
+      create: true
+    ```
+
+    然后升级 TiDB Operator ，具体可参考[升级 TiDB Operator 文档]((upgrade-TiDB Operator.md))。
+
+> **注意：**
+>
+> TiDB Operator 通过开启 `AdvancedStatefulSet` 特性，会将当前 `StatefulSet`
+> 对象转换成 `AdvancedStatefulSet 对象。但并不支持关闭 `AdvancedStatefulSet`
+> 特性后，自动从 `AdvancedStatefulSet` 转换为 Kubernetes 内置的 `StatefulSet`
+> 对象。
+
+## 使用
+
+### 通过 kubectl 查看 AdvancedStatefulSet 对象
+
+`AdvancedStatefulSet` 数据格式与 `StatefulSet` 完全一致，但以 CRD
+方式实现，别名为 `asts` ，可通过一下方法查看命令空间下的对象。
+
+{{< copyable "shell-regular" >}}
+
+```shell
+kubectl get -n <namespace> asts
+```
+
+### 操作 TidbCluster 对象指定 pod 进行缩容
+
+使用增强型 StatefulSet 时，在对 TidbCluster 进行缩容时，除了减少副本数，可同时通过配置 annotations 指定对 pd, tidb 或 tikv 组件下任意一个 pod 进行缩容。
+
+比如：
+
+{{< copyable "" >}}
+
+```yaml
+apiVersion: pingcap.com/v1alpha1
+kind: TidbCluster
+metadata:
+  name: asts
+spec:
+  version: v3.0.12
+  timezone: UTC 
+  pvReclaimPolicy: Delete
+  pd: 
+    baseImage: pingcap/pd
+    replicas: 3
+    requests:
+      storage: "1Gi"
+    config: {}
+  tikv:
+    baseImage: pingcap/tikv
+    replicas: 4
+    requests:
+      storage: "1Gi"
+    config: {}
+  tidb:
+    baseImage: pingcap/tidb
+    replicas: 2
+    service:
+      type: ClusterIP
+    config: {}
+```
+
+会部署 4 个 tikv 实例，分别为 basic-tikv-0, basic-tikv-1, ..., basic-tikv-3。若想缩容掉 basic-tikv-1 需要在修改 `spec.tikv.replicas` 为 3 同时配置以下 annotations:
```

Comment:
Suggest adding scaling out cases:

- scaling out one pod that has been deleted before
- scaling out one pod that has not been deleted before

Reply:
May also include PVC operations.

Reply:
done